concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -465,101 +468,122 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
 
 -- |Add a transaction to the transaction table.  The 'Slot' should be
 -- the slot number of the block that the transaction was received with.
 -- This function should only be called when a transaction is received as part of a block.
 -- The difference from the above function is that this function returns an already existing
 -- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
-          Added bi@WithMetadata{..} -> do
-              ptrs <- getPendingTransactions
-              case wmdData of
-                NormalTransaction tx -> do
-                  focus <- getFocusBlock
-                  st <- blockState focus
-                  macct <- getAccount st $! transactionSender tx
-                  nextNonce <- fromMaybe minNonce <$> mapM (getAccountNonce . snd) macct
-                  -- If a transaction with this nonce has already been run by
-                  -- the focus block, then we do not need to add it to the
-                  -- pending transactions.  Otherwise, we do.
-                  when (nextNonce <= transactionNonce tx) $
-                      putPendingTransactions $! addPendingTransaction nextNonce WithMetadata{wmdData=tx,..} ptrs
-                CredentialDeployment _ ->
-                  putPendingTransactions $! addPendingDeployCredential wmdHash ptrs
-                ChainUpdate cu -> do
-                    focus <- getFocusBlock
-                    st <- blockState focus
-                    nextSN <- getNextUpdateSequenceNumber st (updateType (uiPayload cu))
-                    when (nextSN <= updateSeqNumber (uiHeader cu)) $
-                        putPendingTransactions $! addPendingUpdate nextSN cu ptrs
-              return (Just bi, ResultSuccess)
-          Duplicate tx -> return (Just tx, ResultDuplicate)
-          ObsoleteNonce -> return (Nothing, ResultStale)
+-- This function also verifies the incoming transactions and adds them to the internal
+-- transaction verification cache such that the verification result can be used by the 'Scheduler'.
+-- The @origin@ parameter means if the transaction was received individually or as part of a block.
+-- The function returns the 'BlockItem' if it was "successfully verified" and added to the transaction table.
+-- Note. "Successfully verified" depends on the 'TransactionOrigin', see 'definitelyNotValid' below for details.
+doReceiveTransactionInternal :: (TreeStateMonad pv m) => TV.TransactionOrigin m -> BlockItem -> Timestamp -> Slot -> m (Maybe (BlockItem, TV.VerificationResult), (UpdateResult, TransactionReceiveOutcome))
+doReceiveTransactionInternal origin tr ts slot = do
+   -- Don't accept the transaction if it was expired
+   if transactionExpired (msgExpiry tr) ts then return (Nothing, (ResultStale, Rejected))
```

Comment:
Is there a particular reason why this is still outside here, and not in addCommitTransaction?

Reply:
No. I will move it to the `verify` function of the `TransactionVerification` module.